cctools-2/man/arch.3
.TH ARCH 3 "May 5, 1998" "Apple Computer, Inc."
.SH NAME
NXGetAllArchInfos, NXGetLocalArchInfo, NXGetArchInfoFromName,
NXGetArchInfoFromCpuType, NXFindBestFatArch, NXCombineCpuSubtypes
\- get architecture information
.SH SYNOPSIS
.nf
.PP
.ft B
#include <mach-o/arch.h>
.sp .5
.B extern const NXArchInfo *NXGetAllArchInfos(void);
.sp .5
.B extern const NXArchInfo *NXGetLocalArchInfo(void);
.sp .5
.B extern const NXArchInfo *NXGetArchInfoFromName(const char *\fIname\fB);
.sp .5
.B extern const NXArchInfo *NXGetArchInfoFromCpuType(cpu_type_t \fIcputype\fB,
	\fBcpu_subtype_t \fIcpusubtype\fB);
.sp .5
.B extern struct fat_arch *NXFindBestFatArch(cpu_type_t \fIcputype\fB,
	\fBcpu_subtype_t \fIcpusubtype\fB,
	\fBstruct fat_arch *\fIfat_archs\fB,
	\fBunsigned long \fInfat_archs\fB);
.sp .5
.B extern cpu_subtype_t NXCombineCpuSubtypes(cpu_type_t \fIcputype\fB,
	\fBcpu_subtype_t \fIcpusubtype1\fB,
	\fBcpu_subtype_t \fIcpusubtype2\fB);
.SH DESCRIPTION
These functions are intended for use in programs that have to deal with fat
files or programs that can target multiple architectures.  Typically, a program 
will use a command-line argument that starts with
.BI \-arch " name"\fR,
where this specifies an architecture.  These functions and data structures
provide some help for processing architecture flags and then processing
the contents of a fat file.
.PP
The structure NXArchInfo
is defined in
.I <mach-o/arch.h>
as follows:
.nf
	typedef struct {
		const char *name;
		cpu_type_t cputype;
		cpu_subtype_t cpusubtype;
		enum NXByteOrder byteorder;
		const char *description;
	} NXArchInfo;
.fi
It is used to hold the name of the architecture and the corresponding CPU type
and CPU subtype, together with the architecture's byte order and a brief description string.
.PP
The currently known architectures are:
.sp
.if t .in +0.1i
.if n .ll \n(LLu+1i
.ta +\w'hppa7100LC  'u +\w'CPU_TYPE_POWERPC  'u +\w'CPU_SUBTYPE_MC68030_ONLY  'u +\w'Intel Pentium II Model 5   'u
.br
\fBName	CPU Type	CPU Subtype	Description\fR
.br
pcc	CPU_TYPE_POWERPC	CPU_SUBTYPE_POWERPC_ALL	PowerPC
.br
i386	CPU_TYPE_I386	CPU_SUBTYPE_I386_ALL	Intel 80x86
.br
m68k	CPU_TYPE_MC680x0	CPU_SUBTYPE_MC680x0_ALL	Motorola 68K
.br
hppa	CPU_TYPE_HPPA	CPU_SUBTYPE_HPPA_ALL	HP-PA
.br
i860	CPU_TYPE_I860	CPU_SUBTYPE_I860_ALL	Intel 860
.br
m88k	CPU_TYPE_MC88000	CPU_SUBTYPE_MC88000_ALL	Motorola 88K
.br
sparc	CPU_TYPE_SPARC	CPU_SUBTYPE_SPARC_ALL	SPARC
.br
ppc601	CPU_TYPE_POWERPC	CPU_SUBTYPE_POWERPC_601	PowerPC 601
.br
ppc603	CPU_TYPE_POWERPC	CPU_SUBTYPE_POWERPC_603	PowerPC 603
.br
ppc604	CPU_TYPE_POWERPC	CPU_SUBTYPE_POWERPC_604	PowerPC 604
.br
ppc604e	CPU_TYPE_POWERPC	CPU_SUBTYPE_POWERPC_604e	PowerPC 604e
.br
i486	CPU_TYPE_I386	CPU_SUBTYPE_486	Intel 486
.br
i486SX	CPU_TYPE_I386	CPU_SUBTYPE_486SX	Intel 486SX
.br
pentium	CPU_TYPE_I386	CPU_SUBTYPE_PENT	Intel Pentium
.br
i586	CPU_TYPE_I386	CPU_SUBTYPE_586	Intel 586
.br
pentpro	CPU_TYPE_I386	CPU_SUBTYPE_PENTPRO	Intel Pentium Pro
.br
pentIIm3	CPU_TYPE_I386	CPU_SUBTYPE_PENTII_M3	Intel Pentium II Model 3
.br
pentIIm5	CPU_TYPE_I386	CPU_SUBTYPE_PENTII_M5	Intel Pentium II Model 5
.br
m68030	CPU_TYPE_MC680x0	CPU_SUBTYPE_MC68030_ONLY	Motorola 68030
.br
m68040	CPU_TYPE_MC680x0	CPU_SUBTYPE_MC68040	Motorola 68040
.br
hppa7100LC	CPU_TYPE_HPPA	CPU_SUBTYPE_HPPA_7100LC	HP-PA 7100LC
.if t .in -0.1i
.if n .ll \n(LLu
.PP
The first set of entries
are used for the architecture family.  The second set of entries are used for a specific
architecture, when more than one specific architecture is supported in a
family of architectures.
.PP
.IR NXGetAllArchInfos ()
returns a pointer to an array of all known
NXArchInfo structures.  The last NXArchInfo is marked by a
.SM NULL
name.
.PP
.IR NXGetLocalArchInfo ()
returns the NXArchInfo for the local host, or
.SM NULL
if none is known. 
.PP
.IR NXGetArchInfoFromName ()
and
.IR NXGetArchInfoFromCpuType ()
return the NXArchInfo
from the architecture's name or CPU type/CPU subtype
combination.  A CPU subtype of
.SM CPU_SUBTYPE_MULTIPLE
can be used
to request the most general NXArchInfo
known for the given CPU type.
.SM NULL
is returned if no matching NXArchInfo
can be found.
.PP
.IR NXFindBestFatArch ()
is passed a CPU type and CPU subtype and a set of
fat_arch structs.  It selects the best one that matches (if any), and returns
a pointer to that fat_arch struct (or \s-1NULL\s+1).
The fat_arch structs must be
in the host byte order and correct such that
.I fat_archs 
really points to
enough memory for 
.I nfat_arch 
structs.  It is possible that this routine could
fail if new CPU types or CPU subtypes are added and an old version of this
routine is used.  But if there is an exact match between the CPU type and
CPU subtype and one of the fat_arch structs, this routine will always succeed.
.PP
.IR NXCombineCpuSubtypes ()
returns the resulting CPU subtype when combining two
different CPU subtypes for the specified CPU type.  If the two CPU subtypes
can't be combined (the specific subtypes are mutually exclusive), \-1 is
returned, indicating it is an error to combine them.  This can also fail and
return \-1 if new CPU types or CPU subtypes are added and an old version of
this routine is used.  But if the CPU subtypes are the same, they can always
be combined and this routine will return the CPU subtype passed in.
.\" .SH "SEE ALSO"
.\" .IR arch (1)


objc-1/ChangeLog
Changes from Ali to all the string class files. (12/23/92)
Changes from Derek to deal with name conflicts.

Cleaned up some warnings (7/23/91):

	* Makefile (COMMONFLAGS):  Changed -Wimplicit to -Wall.
	Let's make this compile	without warnings!

	* List.m, Storage.m (read:, write:):  Changed to return self.

	* <various files>:  Removed all unused local variables.

Added support for string objects (7/23/91):

	* NXString.m, NXCharacterSet.m, unichar.h, StringTest.m:  New files.

	* Makefile (LOCAL_CLASSES):  Added NXString.m and NXCharacterSet.m

	* Makefile (LOCAL_INCFILES):  Added unichar.h

	* objc-runtime.m (_objc_fixup_string_object_class):  New routine.
	If NXConstantString is present in class hash table, copy it to
	_NXConstantString_class and insert this address in class hashtable.

	* objc-runtime.m (_objcInit):  Call new routine after classes have
	been fixed up (in _objc_map_selectors).

	* objc-globaldata.m (_NXConstantString_class):  New global variable.
	Reduce pad by 36 bytes to compensate.

	* spec_sys:  Added slots for NXCompareCharacters(), NXEmptyString(),
	and NXFindCharacters().  Added files NXString.o and NXCharacterSet.o
	These changes will need to be integrated into libsys's copy.

Changed -[Object copy] (7/24/91):

	* Object.m (copy):  Chnaged to return [self copyFromZone: [self zone]]
	rather than using _copy.  The function object_copy() is no longer used
	by the runtime system, and could probably be removed.

	* List.[hm] (copy):  Deleted redundant method.

	* Storage.[hm] (copy):  Deleted redundant method.

	* HashTable.[hm] (copy):  Deleted redundant method.

	* NXString.[hm] (copy):  Deleted redundant method.

	* NXCharacterSet.[hm] (copy):  Deleted redundant method.

Changed method names in Storage (7/24/91):

	* Storage.[hm] (replaceElementAt:with:)  Changed from replace:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (insertElement:at:)  Changed from insert:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (removeElementAt:)  Changed from removeAt:
	and added 2.0 compatibility method in implementation only.

Cleaned up Makefile slightly (7/24/91):

	* Makefile (BY_PRODUCTS):  Added objcopt so that make clean removes
	all files not installed by make installsrc.

Fixed bug #14923 in HashTable (7/24/91):

	* HashTable.m (nextState:key:value:):  Set key and value to NULL
	when returning NO.

Fixed bug #14309 in Storage (7/24/91):

	* Storage.m (setNumSlots:):  Changed incorrect test
	(numSlots > maxElements) to (maxElements > old).

	* Storage.m (setNumSlots:):  Changed to use a single realloc.

Fixed bug #5687 in hashtable (7/24/91):

	* hashtable.h (NXHashInsertIfAbsent):  Improved comment.

Added appendList: method (dmoore 7/24/91)

	* List.m, List.h added method.
	
Added maptable functionality (dmoore 7/24/91)

	* maptable.c maptable.m

Removed typedef for id when compiling Objective-C (7/26/91):

	* objc.h (id):  Do not define typedef for id when compiling
	Objective-C (when __OBJC__ is predefined).
	
Made Objective C check for duplicate classes and use the one in the application
	(dmoore 7/27/91)
	
	* objc-runtime.m:  Added following functions:
		_objc_inform:	Just prints message for now.
		_resolveClassConflict: Makes sure that old class goes back in
			if it's from the app.
		_nameForHeader:	Returns the name of an image.
		_getExecHeader:	Picks out the app header.
		_headerForClass: Finds the mach_header containing a class.
		
		Modified _objc_get_classes to check for duplicates.
		
Added maptable prototypes to objc-globaltext.m (dmoore 7/27/91)

Modified spec_sys to include NXString and maptable stuff.

snaroff (7/31/91):

Integrate blaines kernel mods in a rational way.

Renamed "_objc_internal_error" to "_objc_inform"...removed "_objc_inform",
they were nearly identical (snaroff 7/31/91)

Create file "objc-errors.m"...it includes _objc_fatal, _objc_inform,
__objc_error, and _objc_error. MUST ADD _objc_inform and NXLogError to list 
of private externs!!!!!

Call _objc_fatal() instead of perror() in objc-sel.m...callings perror in these situations is bogus.

Have hashtable use NXLogError instead of fprintf(). 

Add field to objc_class for protocols...add protocol templates to objc-class.h

Installed new NXString files from ali (8/3/91):

	* NXString.[hm]:  Most methods are now statically typed.
	
	* NXStringTest.m  Changes to match.

Changes for clean compilation with new compiler (8/3/91):

	* Makefile (objcopt):  Use -ObjC flag with objc-edit.c

	* objcedit.c (round):  Made declaration global as well as definition.

	* objcedit.c (map_input):  Cast third arg to map_fd().

	* objcopt.c (get_objc):  Cast to (SEL *) two places.

	* objc-sel.m (_sel_writeHashTable):  Cast to (POHASH) two places.

	* objc-class.m (cache_flush, _cache_expand, _cache_flush):
	Made iteration variable unsigned.

	* hashtable.m (CopyIntoReadOnly):  Made iteration variable unsigned.

	* typedstream.m (InternalWriteObject):  Cleaned up printf format
	specifiers to match args.

	* typedstreamprivate.h:  Changed all function declarations to static.

	* objc-utils.m:  Cleaned up printf format specifiers to match args.

	* NXString.m (getBytes:range:, getCharacters:range:, NXFindCharacters,
	find:range:flags:usingTable:, _debugPrint):
	Made iteration variable unsigned.

	* NXCharacterSet.m (init, addCharacters:length:,
	removeCharacters:length:):  Made iteration variable unsigned.

Improvements to NXConstantString's (8/12/91):

	* NXString.[hm] (new):  Add version which raises exception.

	* NXString.[hm] (copyFromZone:):  Return an NXReadOnlyString copy
	instead of self, since our code might be unloaded.

	* NXString.[hm] (write:):  Write an NXReadOnlyString copy.

	* NXString.[hm] (finishUnarchiving):  Removed.

	* NXString.[hm] (free):  This now raises an error, since copyFromZone:
	no longer returns self.  You never allocate NXConstantString's, so you
	should never free them.


dpkg-3/dpkg.src/lib/arch.c
#include "arch.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <sys/utsname.h>

#include "varbuf.h"
#include "dpkg.h"
#include "dpkg-db.h"
#include "libdpkg-int.h"
#include "config.h"

#define DPKG_ARCHSTR_MAX 256

static char const *archcompat[] = 
{ "i386", "sparc", "alpha", "m68k", "arm", "powerpc" };

int dpkg_compatible_architecture (char *hostarch, char *pkgarch)
{
  return 0;
}

/* destroys buf */

int dpkg_parse_architecture (char *buf, char **cpu, char **manufacturer, char **kernel, char **os)
{
  char *s = buf;
  *cpu = *manufacturer = *kernel = *os = NULL;

  *cpu = s;
  s = index (s, '-');
  if (s == NULL) { *cpu = NULL; return -1; }
  *s++ = '\0';
  
  *manufacturer = s;
  s = index (s, '-');
  if (s == NULL) { *cpu = *manufacturer = NULL; return -1; }
  *s++ = '\0';
  
  *os = s;
  s = index (s, '-');
  if (s == NULL) { return 0; }
  *s++ = '\0';

  /* if we get here, it's CPU-MANUFACTURER-KERNEL-OS, not CPU-MANUFACTURER-OS */

  *kernel = *os;
  *os = s;
  return 0;
}

int dpkg_canonicalize_architecture (const char *src, unsigned char *dst, size_t n)
{
  const char *cpu, *manufacturer, *kernel, *os;
  char buf[DPKG_ARCHSTR_MAX];
  int ret;
  
  ret = snprintf (buf, DPKG_ARCHSTR_MAX, "%s", src);
  if (ret < 0) { return ret; }
  
  ret = dpkg_parse_architecture
    (buf, (char **) &cpu, (char **) &manufacturer, (char **) &kernel, (char **) &os);
  if (ret != 0) { ohshit (_("unable to parse architecture string `%.250s'"), src); }
  
  if ((cpu == NULL) || (manufacturer == NULL) || (os == NULL)) { return -2; }
  
  if ((kernel != NULL) && (strcmp (kernel, "linux") == 0)) {
    /* remap common processor aliases */
    if ((strcmp (cpu, "i486") == 0)
	|| (strcmp (cpu, "i586") == 0)
	|| (strcmp (cpu, "i686") == 0)
	|| (strcmp (cpu, "pentium") == 0)) { 
      cpu = "i386";
    } else if (strcmp (cpu, "ppc") == 0) {
      cpu = "powerpc";
    }

    if (strcmp (manufacturer, "pc") == 0) {
      manufacturer = "unknown";
    }
  }

  if (strncmp (os, "rhapsody", strlen ("rhapsody")) == 0) {
    os = "rhapsody";
  } else if (strncmp (os, "rhapsody", strlen ("rhapsody")) == 0) {
    os = "darwin";
  } else {
    os = "unknown";
  }

  if (kernel != NULL) {
    return snprintf (dst, n, "%s-%s-%s-%s", cpu, manufacturer, kernel, os);
  } else {
    return snprintf (dst, n, "%s-%s-%s", cpu, manufacturer, os);
  }
}

int dpkg_expand_archcompat (const char *src, char *dst, size_t n)
{
  const char **sptr = NULL;
  for (sptr = archcompat; (*sptr) != NULL; sptr++) {
    if (strcmp (*sptr, src) == 0) {
      return snprintf (dst, n, "%s-unknown-linux-gnu", src);
    }
  }
  return snprintf (dst, n, "%s", src);
}

int dpkg_reduce_archcompat (const char *src, char *dst, size_t n)
{
  const char **sptr = NULL;
  for (sptr = archcompat; (*sptr) != NULL; sptr++) {
    size_t len = strlen (*sptr);
    if ((strncmp (*sptr, src, len) == 0) 
	&& ((strcmp ((src + len), "-unknown-linux-gnu") == 0)
	    || (strcmp ((src + len), "-pc-linux-gnu") == 0))) {
      return snprintf (dst, n, *sptr);
    }
  }
  return snprintf (dst, n, "%s", src);
}

/* return the GNU-format architecture string of the machine on which
   this version of dpkg was compiled */

int dpkg_find_gnu_build_architecture (char *dst, size_t n)
{
  return snprintf (dst, n, "%s", ARCHITECTURE);
}

int dpkg_find_gnu_installation_architecture (unsigned char *str, size_t n)
{
  int ret;
  unsigned char buf[DPKG_ARCHSTR_MAX];

  ret = dpkg_find_gnu_build_architecture (buf, DPKG_ARCHSTR_MAX);
  if (ret <= 0) { return ret; }
  ret = dpkg_canonicalize_architecture (buf, str, n);
  return ret;
}

static void badlgccfn (const char *compiler, const char *output, const char *why) NONRETURNING;
static void badlgccfn (const char *compiler, const char *output, const char *why) {
  fprintf (stderr,
	   _("dpkg: unexpected output from `%s --print-libgcc-file-name':\n"
	     " `%s'\n"),
	   compiler,output);
  ohshit(_("compiler libgcc filename not understood: %.250s"),why);
}

int dpkg_find_gnu_target_architecture (char *dst, size_t n)
{
  struct utsname uts;
  
  if (uname (&uts) < 0) {
    ohshite (_("unable to determine system information using uname()"));
  }

  /* for the moment, we support only native compilation on Rhapsody */
  if (strcmp (uts.sysname, "Rhapsody") == 0) {
    return snprintf (dst, n, "%s-%s", uts.machine, "apple-rhapsody");
  } else if (strcmp (uts.sysname, "Darwin") == 0) {
    return snprintf (dst, n, "%s", "powerpc-apple-darwin");
  }

  /* on linux, use gcc --print-libgcc-file-name to find the compilation architecture */
  if (strcmp (uts.sysname, "Linux") == 0) {

    const char *ccompiler;
    int p1[2], c;
    pid_t c1;
    FILE *ccpipe;
    struct varbuf vb;
    char *p, *q;

    ccompiler = getenv ("CC");
    if (ccompiler == NULL) { ccompiler= "gcc"; }
    varbufinit (&vb);
    m_pipe (p1);
    ccpipe = fdopen (p1[0], "r"); 
    if (! ccpipe) { ohshite (_("failed to fdopen CC pipe")); }
    if (! (c1 = m_fork ())) {
      m_dup2 (p1[1], 1); 
      close (p1[0]); 
      close (p1[1]);
      execlp (ccompiler, ccompiler, "--print-libgcc-file-name", (char*) 0);
      ohshite (_("failed to exec C compiler `%.250s'"), ccompiler);
    }
    close (p1[1]);

    while ((c = getc (ccpipe)) != EOF) { varbufaddc (&vb,c); }
    if (ferror(ccpipe)) ohshite(_("error reading from CC pipe"));
    waitsubproc (c1, "gcc --print-libgcc-file-name", 0);
    if (!vb.used) { badlgccfn (ccompiler, "", _("empty output")); }
    varbufaddc (&vb,0);

    if (vb.buf[vb.used-2] != '\n') { badlgccfn (ccompiler, vb.buf, _("no newline")); }
    vb.used -= 2;
    varbufaddc (&vb,0);
    p = strstr (vb.buf, "/gcc-lib/");
    if (p == NULL) { badlgccfn (ccompiler, vb.buf, _("no gcc-lib component")); }
    p += 9;
    q = strchr (p, '-'); 
    if (q == NULL) { badlgccfn (ccompiler, vb.buf, _("no hyphen after gcc-lib")); }
    q = strchr (q, '/');
    if (q == NULL) { badlgccfn (ccompiler, vb.buf, _("no / after arch-directory")); }
    *q = '\0';
    q = strchr (p, '-');
    if (q != NULL) { *q = '\0'; }

    return snprintf (dst, n, "%s-unknown-linux-gnu", p);
  }

  return 0;
}
